File: golang/3151_is_array_special/is_array_special.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func isArraySpecial(nums []int) bool {
	n := len(nums)
	for i := 1; i < n; i++ {
		if nums[i-1]%2 == nums[i]%2 {
			return false
		}
	}
	return true
}

func main() {
	tests := []struct {
		nums []int
		ans  bool
	}{
		{[]int{1}, true},
		{[]int{2, 1, 4}, true},
		{[]int{4, 3, 1, 6}, false},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, isArraySpecial(test.nums), index)
	}
}

File: cpp/3151_is_array_special/is_array_special.cpp
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  bool isArraySpecial(vector<int>& nums) {
    int n = nums.size();
    for (int i = 1; i < n; ++i) {
      if (nums[i - 1] % 2 == nums[i] % 2) {
        return false;
      }
    }
    return true;
  }
};

int main() {
  vector<tuple<vector<int>, bool>> tests{
      {{1}, true},
      {{2, 1, 4}, true},
      {{4, 3, 1, 6}, false},
  };

  for (auto& [nums, ans] : tests) {
    assert(Solution().isArraySpecial(nums) == ans);
  }
}